pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,397 @@
+		}
+	}
+
+	if overallErr != nil {
+		if revErr := revertStack.Revert(); revErr != nil {
+			log.WithError(revErr).Error("Failed to revert ENI full datapath migration")
+		}
+
+		return overallErr, revertErr
+	}
+
+	for _, r := range rules {
+		switch r.Priority {
+		case linux_defaults.RulePriorityEgress:
+			scopedLog := log.WithField("rule", r)
+
+			routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+				Table: r.Table,
+			}, netlink.RT_FILTER_TABLE)
+			if err != nil {
+				overallErr = fmt.Errorf("failed to list rules routes: %w", err)
+				scopedLog.WithError(overallErr).
+					Debug("Skipping migration of egress rule due to failure in listing routes")
+				break
+			}
+
+			// If there are no routes under the same table as the rule, we will
+			// simply attempt to delete the old rule. This deletion is
+			// best-effort, meaning we won't stop this migration if it fails,
+			// because leaving this rule on the system will not impact
+			// connectivity because there are no routes under it.
+			if len(routes) == 0 {
+				scopedLog.Debug("Skipping migration of egress rule due to no routes found, proceeding to delete rule")
+
+				// As explained in the above comment, this is best-effort so we
+				// ignore the revert stack return value.
+				_, err := deleteOldRule(r)
+				if err != nil {
+					overallErr = err
+					scopedLog.WithError(overallErr).Warning("Failed to cleanup old ENI rule without routes")
+				}
+
+				continue
+			}
+
+			oldTable := r.Table
+
+			revert, err := onUpgradeDeleteRuleAndRoutes(r, routes, oldTable)
+			revertStack.Extend(revert)
+			if err != nil {
+				overallErr = fmt.Errorf("failed to remove old ENI datapath: %w", err)
+				scopedLog.WithError(overallErr).Error("Failed to remove old ENI datapath")
+
+				// The reverting will be outside of this loop since we are
+				// breaking out of the loop.
+				break
+			}
+
+		case linux_defaults.RulePriorityEgressv2:
+			// Downgrade case. Unsupported for now.
+		}
+	}
+
+	if overallErr != nil {
+		if revErr := revertStack.Revert(); revErr != nil {
+			log.WithError(revErr).Error("Failed to revert ENI full datapath migration")
+		}
+
+		return overallErr, revertErr
+	}
+
+	return nil, nil
+}
+
+// retrieveTableIDFromIfIndex computes the correct table ID based on the
+// ifindex provided. The table ID is comprised of the number associated with an
+// ENI device that corresponds to the ifindex, plus the specific table offset
+// value.
+func retrieveTableIDFromIfIndex(ifindex int, getter interfaceDB) (int, error) {
+	link, err := netlink.LinkByIndex(ifindex)
+	if err != nil {
+		return -1, fmt.Errorf("failed to find link by index: %w", err)
+	}
+
+	mac := link.Attrs().HardwareAddr.String()
+	ifaceNum, err := getter.GetInterfaceNumberByMAC(mac)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get interface number by MAC %q: %w", mac, err)
+	}
+
+	return linux_defaults.RouteTableInterfacesOffset + ifaceNum, nil
```

Comment:
What happens if the ifaceNum is e.g. 245? Didn't you mean to to multiple instead?